crates/env/src/contract.rs

```diff
@@ -0,0 +1,134 @@
+/// Stores the used host environment type of the ink! smart contract.
+///
+/// # Note
+///
+/// The used host environment can be altered using the `env` configuration
+/// parameter in the `#[ink::contract]` parameters. For example if the user
+/// wanted to use an environment type definition called `MyEnvironment` they
+/// issue the ink! smart contract as follows:
+///
+/// ```no_compile
+/// #[ink::contract(env = MyEnvironment)]
+/// ```
+///
+/// # Usage: Default Environment
+///
+/// ```
+/// 
+/// #[ink::contract]
+/// pub mod contract {
+///     #[ink(storage)]
+///     pub struct Contract {}
+///
+///     impl Contract {
+///         #[ink(constructor)]
+///         pub fn constructor() -> Self { Self {} }
+///
+///         #[ink(message)]
+///         pub fn message(&self) {}
+///     }
+/// }
+///
+/// use contract::Contract;
+///
+/// # use ink::env::ContractEnv;
+/// # use ink::codegen::utils::IsSameType;
+///
+/// // The following line only compiles successfully if both
+/// // `ink_env::DefaultEnvironment` and `<Contract as ContractEnv>::Env`
+/// // are of the same type.
+/// const _: IsSameType<<Contract as ContractEnv>::Env> =
+///     <IsSameType<ink_env::DefaultEnvironment>>::new();
+/// ```
+///
+/// # Usage: Custom Environment
+///
+/// ```
+/// # use ink_env::{Environment, DefaultEnvironment};
+///
+/// pub struct CustomEnvironment {}
+///
+/// impl Environment for CustomEnvironment {
+///     const MAX_EVENT_TOPICS: usize = 4;
+///
+///     type AccountId = <DefaultEnvironment as Environment>::AccountId;
+///     type Balance = u64;
+///     type Hash = <DefaultEnvironment as Environment>::Hash;
+///     type BlockNumber = u32;
+///     type Timestamp = u64;
+///     type ChainExtension = <DefaultEnvironment as Environment>::ChainExtension;
+/// }
+///
+/// #[ink::contract(env = super::CustomEnvironment)]
+/// pub mod contract {
+///     #[ink(storage)]
+///     pub struct Contract {}
+///
+///     impl Contract {
+///         #[ink(constructor)]
+///         pub fn constructor() -> Self { Self {} }
+///
+///         #[ink(message)]
+///         pub fn message(&self) {}
+///     }
+/// }
+///
+/// use contract::Contract;
+/// # use ink::env::ContractEnv;
+/// # use ink::codegen::utils::IsSameType;
+///
+/// // The following line only compiles successfully if both
+/// // `CustomEnvironment` and `<Contract as ContractEnv>::Env`
+/// // are of the same type.
+/// const _: IsSameType<<Contract as ContractEnv>::Env> =
+///     <IsSameType<CustomEnvironment>>::new();
+///
+/// fn main() {}
+/// ```
+pub trait ContractEnv {
```

Comment:
These traits have just been copied from `ink::reflect` so we can use them here in `ink_env`